<a href="https://colab.research.google.com/github/carlosfmorenog/CM4125/blob/main/CM4125_T5/CM4125_Lec5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# CM4125 Topic 5 - Data Loading and Exploration

## Lecture objectives

1) Understand how we define data (for the purpose of this module) and where to get it

2) Learn different methods to import/export data into practical representations

3) Apply basic commands and concepts to explore it

## What is data (for data viz purposes)?

Information or characteristics collected through observation or experimentation

Using a tabular approach...

Each entry/observation corresponds to a `row`

Each feature/characteristic corresponds to a `column`

## Where to get data from?

### "From a friend"


We are given access to a file or register containing the data

There are many formats, but mostly we will deal with:

    .csv
    .tsv
    .txt
    .xlsx
    .json

In [ ]:
## Example: Reading a .csv file into python
## This file contains the age and height of some participants
import pandas as pd
df = pd.read_csv('https://www.dropbox.com/scl/fi/rjxj1fyvyt7ngwrim89d8/data.csv?rlkey=436xq62tyk248xxssneuzzkmu&raw=1')
df

Col 1  Col 2  Col 3
0       Nick     21   1.85
1      Chris     29   1.79
2        Tim     28   1.75
3        Ron     34   1.81
4     Monica     35   1.69
5  Cassandra     21   1.66

### Online repos

* There is a huge amount of free data out there!
* Just be careful where you get it!
* Websites such as [Kaggle](https://www.kaggle.com), [UCI](https://archive.ics.uci.edu/ml/index.php) or [Keel](https://sci2s.ugr.es/keel/datasets.php) contain thousands of examples that you can download and import to your preferred tool
* Sometimes they even contain ways to connect to your data in faster and more secure ways

![Fig. 2](https://www.dropbox.com/s/t2klwwig3qj2dh2/fig2.jpg?raw=1)

### Modules and packages

Languages such as Python and R already contain some preloaded data repositories

They are sometimes in weird shapes and forms

Nonetheless, they are good to start experimenting!

In [ ]:
## Loading the IRIS dataset from the SCIKIT LEARN module
from sklearn.datasets import load_iris
iris = load_iris()
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

## Exploring Data by Rows

Now that we have data in a tabular form, let's see how to access certain positions

To do so, let's use a larger dataset of Netflix Original series contained in a .csv file called `netflix.csv`

In [ ]:
netflix = pd.read_csv('https://www.dropbox.com/scl/fi/xo9zepvclpfe76fzwcwlg/netflix.csv?rlkey=g7q3k91p0lzz2jkznr7hid2fc&raw=1')
netflix

Title                         Genre  \
0                   House of Cards               Political drama   
1                    Hemlock Grove               Horror/thriller   
2          Orange Is the New Black                  Comedy-drama   
3                       Marco Polo              Historical drama   
4                        Bloodline                      Thriller   
..                             ...                           ...   
353                        Busted!  Korean language variety show   
354   The Break with Michelle Wolf                    Late-night   
355      Norm Macdonald Has a Show                     Talk show   
356  Patriot Act with Hasan Minhaj                     Talk show   
357                        The Fix                    Panel show   

                      GenreLabels   Premiere                 Seasons  \
0                 political,drama   1-Feb-13  6 seasons, 73 episodes   
1                 horror,thriller  19-Apr-13  3 seasons, 33 episodes   
2                    comedy-drama  11-Jul-13  6 seasons, 78 episodes   
3                historical,drama  12-Dec-14  2 seasons, 20 episodes   
4                        thriller  20-Mar-15  3 seasons, 33 episodes   
..                            ...        ...                     ...   
353  korean,language,variety,show   4-May-18   1 season, 10 episodes   
354                    late-night  27-May-18   1 season, 10 episodes   
355                     talk,show  14-Sep-18   1 season, 10 episodes   
356                     talk,show  28-Oct-18  3 volumes, 19 episodes   
357                    panel,show  14-Dec-18   1 season, 10 episodes   

     SeasonsParsed  EpisodesParsed      Length  MinLength  MaxLength   Status  \
0                6              73  42–59 min.         42         59    Ended   
1                3              33  45–58 min.         45         58    Ended   
2                6              78  50–92 min.         50         92  Renewed   
3                2              20  48–65 min.         48         65    Ended   
4                3              33  48–68 min.         48         68    Ended   
..             ...             ...         ...        ...        ...      ...   
353              1              10  75–98 min.         75         98  Renewed   
354              1              10     27 min.         27         27  Renewed   
355              1              10  26–35 min.         26         35  Pending   
356              0              19  23–30 min.         23         30  Renewed   
357              1              10  23–31 min.         23         31  Pending   

     Active        Table Language  
0         0        Drama  English  
1         0        Drama  English  
2         1        Drama  English  
3         0        Drama  English  
4         0        Drama  English  
..      ...          ...      ...  
353       1  VarietyTalk   Korean  
354       1  VarietyTalk  English  
355       1  VarietyTalk  English  
356       1  VarietyTalk  English  
357       1  VarietyTalk  English  

[358 rows x 14 columns]

Notice that when the `DataFrame` is shown in Jupyter, it displays `358 rows × 14 columns` at the bottom to tell us how large the dataset is.

### Using a Non-Numerical Index

We can actually use any of the unique-entry columns (for example `Title`, which is column `0`)

In [ ]:
# reloading the dataset, now stating the first column as the index one
netflix = pd.read_csv('https://www.dropbox.com/scl/fi/xo9zepvclpfe76fzwcwlg/netflix.csv?rlkey=g7q3k91p0lzz2jkznr7hid2fc&raw=1',
                      index_col=0)
netflix

Genre  \
Title                                                         
House of Cards                              Political drama   
Hemlock Grove                               Horror/thriller   
Orange Is the New Black                        Comedy-drama   
Marco Polo                                 Historical drama   
Bloodline                                          Thriller   
...                                                     ...   
Busted!                        Korean language variety show   
The Break with Michelle Wolf                     Late-night   
Norm Macdonald Has a Show                         Talk show   
Patriot Act with Hasan Minhaj                     Talk show   
The Fix                                          Panel show   

                                                GenreLabels   Premiere  \
Title                                                                    
House of Cards                              political,drama   1-Feb-13   
Hemlock Grove                               horror,thriller  19-Apr-13   
Orange Is the New Black                        comedy-drama  11-Jul-13   
Marco Polo                                 historical,drama  12-Dec-14   
Bloodline                                          thriller  20-Mar-15   
...                                                     ...        ...   
Busted!                        korean,language,variety,show   4-May-18   
The Break with Michelle Wolf                     late-night  27-May-18   
Norm Macdonald Has a Show                         talk,show  14-Sep-18   
Patriot Act with Hasan Minhaj                     talk,show  28-Oct-18   
The Fix                                          panel,show  14-Dec-18   

                                              Seasons  SeasonsParsed  \
Title                                                                  
House of Cards                 6 seasons, 73 episodes              6   
Hemlock Grove                  3 seasons, 33 episodes              3   
Orange Is the New Black        6 seasons, 78 episodes              6   
Marco Polo                     2 seasons, 20 episodes              2   
Bloodline                      3 seasons, 33 episodes              3   
...                                               ...            ...   
Busted!                         1 season, 10 episodes              1   
The Break with Michelle Wolf    1 season, 10 episodes              1   
Norm Macdonald Has a Show       1 season, 10 episodes              1   
Patriot Act with Hasan Minhaj  3 volumes, 19 episodes              0   
The Fix                         1 season, 10 episodes              1   

                               EpisodesParsed      Length  MinLength  \
Title                                                                  
House of Cards                             73  42–59 min.         42   
Hemlock Grove                              33  45–58 min.         45   
Orange Is the New Black                    78  50–92 min.         50   
Marco Polo                                 20  48–65 min.         48   
Bloodline                                  33  48–68 min.         48   
...                                       ...         ...        ...   
Busted!                                    10  75–98 min.         75   
The Break with Michelle Wolf               10     27 min.         27   
Norm Macdonald Has a Show                  10  26–35 min.         26   
Patriot Act with Hasan Minhaj              19  23–30 min.         23   
The Fix                                    10  23–31 min.         23   

                               MaxLength   Status  Active        Table  \
Title                                                                    
House of Cards                        59    Ended       0        Drama   
Hemlock Grove                         58    Ended       0        Drama   
Orange Is the New Black               92  Renewed       1        Drama   
Marco Polo                 

It's up to you to decide whether to use default numerical indexing or to use an uniquely-identifying column from your dataset.

### Locating an item

In [ ]:
# Locating all info of an entry
netflix.loc['Orange Is the New Black']

Genre                       Comedy-drama
GenreLabels                 comedy-drama
Premiere                       11-Jul-13
Seasons           6 seasons, 78 episodes
SeasonsParsed                          6
EpisodesParsed                        78
Length                        50–92 min.
MinLength                             50
MaxLength                             92
Status                           Renewed
Active                                 1
Table                              Drama
Language                         English
Name: Orange Is the New Black, dtype: object

In [ ]:
# Locating particular info of that entry
netflix.at['Orange Is the New Black', 'Seasons']

'6 seasons, 78 episodes'

In [ ]:
# This one allows you to access by number, even if the table has no numerical indexes
netflix.iloc[0]

Genre                    Political drama
GenreLabels              political,drama
Premiere                        1-Feb-13
Seasons           6 seasons, 73 episodes
SeasonsParsed                          6
EpisodesParsed                        73
Length                        42–59 min.
MinLength                             42
MaxLength                             59
Status                             Ended
Active                                 0
Table                              Drama
Language                         English
Name: House of Cards, dtype: object

### Getting a Subset of the Data

In [ ]:
netflix.head(5)

Genre       GenreLabels   Premiere  \
Title                                                                    
House of Cards            Political drama   political,drama   1-Feb-13   
Hemlock Grove             Horror/thriller   horror,thriller  19-Apr-13   
Orange Is the New Black      Comedy-drama      comedy-drama  11-Jul-13   
Marco Polo               Historical drama  historical,drama  12-Dec-14   
Bloodline                        Thriller          thriller  20-Mar-15   

                                        Seasons  SeasonsParsed  \
Title                                                            
House of Cards           6 seasons, 73 episodes              6   
Hemlock Grove            3 seasons, 33 episodes              3   
Orange Is the New Black  6 seasons, 78 episodes              6   
Marco Polo               2 seasons, 20 episodes              2   
Bloodline                3 seasons, 33 episodes              3   

                         EpisodesParsed      Length  MinLength  MaxLength  \
Title                                                                       
House of Cards                       73  42–59 min.         42         59   
Hemlock Grove                        33  45–58 min.         45         58   
Orange Is the New Black              78  50–92 min.         50         92   
Marco Polo                           20  48–65 min.         48         65   
Bloodline                            33  48–68 min.         48         68   

                          Status  Active  Table Language  
Title                                                     
House of Cards             Ended       0  Drama  English  
Hemlock Grove              Ended       0  Drama  English  
Orange Is the New Black  Renewed       1  Drama  English  
Marco Polo                 Ended       0  Drama  English  
Bloodline                  Ended       0  Drama  English

In [ ]:
netflix.tail(3)

Genre GenreLabels   Premiere  \
Title                                                              
Norm Macdonald Has a Show       Talk show   talk,show  14-Sep-18   
Patriot Act with Hasan Minhaj   Talk show   talk,show  28-Oct-18   
The Fix                        Panel show  panel,show  14-Dec-18   

                                              Seasons  SeasonsParsed  \
Title                                                                  
Norm Macdonald Has a Show       1 season, 10 episodes              1   
Patriot Act with Hasan Minhaj  3 volumes, 19 episodes              0   
The Fix                         1 season, 10 episodes              1   

                               EpisodesParsed      Length  MinLength  \
Title                                                                  
Norm Macdonald Has a Show                  10  26–35 min.         26   
Patriot Act with Hasan Minhaj              19  23–30 min.         23   
The Fix                                    10  23–31 min.         23   

                               MaxLength   Status  Active        Table  \
Title                                                                    
Norm Macdonald Has a Show             35  Pending       1  VarietyTalk   
Patriot Act with Hasan Minhaj         30  Renewed       1  VarietyTalk   
The Fix                               31  Pending       1  VarietyTalk   

                              Language  
Title                                   
Norm Macdonald Has a Show      English  
Patriot Act with Hasan Minhaj  English  
The Fix                        English

In [ ]:
netflix.sample(5)

Genre          GenreLabels  \
Title                                                                        
My First First Love                   Romantic comedy      romantic,comedy   
Beyond Stranger Things                      Aftershow            aftershow   
Larva Island                      childrens-animation  childrens-animation   
Trigger Warning with Killer Mike          Docu-series          docu-series   
Cooking on High                          Cooking show         cooking,show   

                                   Premiere                 Seasons  \
Title                                                                 
My First First Love               18-Apr-19    1 season, 8 episodes   
Beyond Stranger Things            27-Oct-17    1 season, 7 episodes   
Larva Island                      19-Oct-18  2 seasons, 26 episodes   
Trigger Warning with Killer Mike  18-Jan-19    1 season, 6 episodes   
Cooking on High                   22-Jun-18   1 season, 12 episodes   

                                  SeasonsParsed  EpisodesParsed      Length  \
Title                                                                         
My First First Love                           1               8  46–56 min.   
Beyond Stranger Things                        1               7  15–25 min.   
Larva Island                                  2              26      8 min.   
Trigger Warning with Killer Mike              1               6  22–27 min.   
Cooking on High                               1              12  13–15 min.   

                                  MinLength  MaxLength   Status  Active  \
Title                                                                     
My First First Love                      46         56  Renewed       1   
Beyond Stranger Things                   15         25  Pending       1   
Larva Island                              8          8  Pending       1   
Trigger Warning with Killer Mike         22         27  Pending       1   
Cooking on High                          13         15    Ended       0   

                                            Table Language  
Title                                                       
My First First Love               ForeignLanguage   Korean  
Beyond Stranger Things                VarietyTalk  English  
Larva Island                             Children  English  
Trigger Warning with Killer Mike       DocuSeries  English  
Cooking on High                           Reality  English

### Choosing Specific Rows

In [ ]:
netflix.loc[['Diablero', 'Motown Magic', 'Typewriter']] # by index

Genre              GenreLabels   Premiere  \
Title                                                                       
Diablero      Horror fantasy thriller  horror,fantasy,thriller  21-Dec-18   
Motown Magic      childrens-animation      childrens-animation  20-Nov-18   
Typewriter                     Horror                   horror  19-Jul-19   

                            Seasons  SeasonsParsed  EpisodesParsed  \
Title                                                                
Diablero       1 season, 8 episodes              1               8   
Motown Magic  1 season, 25 episodes              1              25   
Typewriter                      TBA              0               0   

                  Length  MinLength  MaxLength   Status  Active  \
Title                                                             
Diablero      36–44 min.         36         44  Renewed       1   
Motown Magic  15–26 min.         15         26  Renewed       1   
Typewriter           TBA          0          0  Pending       1   

                        Table Language  
Title                                   
Diablero      ForeignLanguage  Spanish  
Motown Magic         Children  English  
Typewriter    ForeignLanguage    Hindi

In [ ]:
netflix.iloc[[40, 12, 106, 79]] # by numerical index

Genre  \
Title                                                             
Jessica Jones                   Neo-noir/psychological thriller   
A Series of Unfortunate Events             Black-comedy mystery   
Lost Song                                       Musical fantasy   
After Life                                               Comedy   

                                                    GenreLabels   Premiere  \
Title                                                                        
Jessica Jones                   neo-noir,psychological,thriller  20-Nov-15   
A Series of Unfortunate Events             black-comedy,mystery  13-Jan-17   
Lost Song                                       musical,fantasy  31-Mar-18   
After Life                                               comedy   8-Mar-19   

                                               Seasons  SeasonsParsed  \
Title                                                                   
Jessica Jones                   3 seasons, 39 episodes              3   
A Series of Unfortunate Events  3 seasons, 25 episodes              3   
Lost Song                        1 season, 12 episodes              1   
After Life                        1 season, 6 episodes              1   

                                EpisodesParsed      Length  MinLength  \
Title                                                                   
Jessica Jones                               39  44–56 min.         44   
A Series of Unfortunate Events              25  36–64 min.         36   
Lost Song                                   12  23–24 min.         23   
After Life                                   6  25–31 min.         25   

                                MaxLength   Status  Active   Table Language  
Title                                                                        
Jessica Jones                          56    Ended       0  Marvel  English  
A Series of Unfortunate Events         64    Ended       0   Drama  English  
Lost Song                              24    Ended       0   Anime  English  
After Life                             31  Renewed       1  Comedy  English

In [ ]:
netflix.loc['Maniac':'After Life'] # selecting a range by index

Genre                 GenreLabels  \
Title                                                                         
Maniac                              Dark comedy                 dark,comedy   
The Kominsky Method                      Comedy                      comedy   
Sex Education        Coming-of-age comedy-drama  coming-of-age,comedy-drama   
Russian Doll                             Comedy                      comedy   
After Life                               Comedy                      comedy   

                      Premiere               Seasons  SeasonsParsed  \
Title                                                                 
Maniac               21-Sep-18           10 episodes              0   
The Kominsky Method  16-Nov-18  1 season, 8 episodes              1   
Sex Education        11-Jan-19  1 season, 8 episodes              1   
Russian Doll          1-Feb-19  1 season, 8 episodes              1   
After Life            8-Mar-19  1 season, 6 episodes              1   

                     EpisodesParsed      Length  MinLength  MaxLength  \
Title                                                                   
Maniac                           10  27–47 min.         27         47   
The Kominsky Method               8  23–34 min.         23         34   
Sex Education                     8  47–53 min.         47         53   
Russian Doll                      8  25–30 min.         25         30   
After Life                        6  25–31 min.         25         31   

                         Status  Active   Table Language  
Title                                                     
Maniac               Miniseries       0  Comedy  English  
The Kominsky Method     Renewed       1  Comedy  English  
Sex Education           Renewed       1  Comedy  English  
Russian Doll            Renewed       1  Comedy  English  
After Life              Renewed       1  Comedy  English

In [ ]:
netflix.iloc[75:79] # selecting a range by numerical index

Genre                 GenreLabels  \
Title                                                                         
Maniac                              Dark comedy                 dark,comedy   
The Kominsky Method                      Comedy                      comedy   
Sex Education        Coming-of-age comedy-drama  coming-of-age,comedy-drama   
Russian Doll                             Comedy                      comedy   

                      Premiere               Seasons  SeasonsParsed  \
Title                                                                 
Maniac               21-Sep-18           10 episodes              0   
The Kominsky Method  16-Nov-18  1 season, 8 episodes              1   
Sex Education        11-Jan-19  1 season, 8 episodes              1   
Russian Doll          1-Feb-19  1 season, 8 episodes              1   

                     EpisodesParsed      Length  MinLength  MaxLength  \
Title                                                                   
Maniac                           10  27–47 min.         27         47   
The Kominsky Method               8  23–34 min.         23         34   
Sex Education                     8  47–53 min.         47         53   
Russian Doll                      8  25–30 min.         25         30   

                         Status  Active   Table Language  
Title                                                     
Maniac               Miniseries       0  Comedy  English  
The Kominsky Method     Renewed       1  Comedy  English  
Sex Education           Renewed       1  Comedy  English  
Russian Doll            Renewed       1  Comedy  English

### Filtering Rows by Values in Columns

One of the most important features you need is to be able to filter observations out

We can define a **condition** so that it can be checked for all entries

In [ ]:
our_condition = netflix['SeasonsParsed'] == 6
our_condition

Title
House of Cards                    True
Hemlock Grove                    False
Orange Is the New Black           True
Marco Polo                       False
Bloodline                        False
                                 ...  
Busted!                          False
The Break with Michelle Wolf     False
Norm Macdonald Has a Show        False
Patriot Act with Hasan Minhaj    False
The Fix                          False
Name: SeasonsParsed, Length: 358, dtype: bool

Then, we can ask for that condition to be the filter of the original dataset

In [ ]:
netflix[our_condition]

Genre          GenreLabels  \
Title                                                                       
House of Cards                       Political drama      political,drama   
Orange Is the New Black                 Comedy-drama         comedy-drama   
The Adventures of Puss in Boots  childrens-animation  childrens-animation   
Dragons: Race to the Edge        childrens-animation  childrens-animation   
Trolls: The Beat Goes On!        childrens-animation  childrens-animation   

                                  Premiere                 Seasons  \
Title                                                                
House of Cards                    1-Feb-13  6 seasons, 73 episodes   
Orange Is the New Black          11-Jul-13  6 seasons, 78 episodes   
The Adventures of Puss in Boots  16-Jan-15  6 seasons, 78 episodes   
Dragons: Race to the Edge        26-Jun-15  6 seasons, 78 episodes   
Trolls: The Beat Goes On!        19-Jan-18  6 seasons, 38 episodes   

                                 SeasonsParsed  EpisodesParsed      Length  \
Title                                                                        
House of Cards                               6              73  42–59 min.   
Orange Is the New Black                      6              78  50–92 min.   
The Adventures of Puss in Boots              6              78  22–24 min.   
Dragons: Race to the Edge                    6              78  22–23 min.   
Trolls: The Beat Goes On!                    6              38  24–25 min.   

                                 MinLength  MaxLength   Status  Active  \
Title                                                                    
House of Cards                          42         59    Ended       0   
Orange Is the New Black                 50         92  Renewed       1   
The Adventures of Puss in Boots         22         24    Ended       0   
Dragons: Race to the Edge               22         23    Ended       0   
Trolls: The Beat Goes On!               24         25  Pending       1   

                                    Table Language  
Title                                               
House of Cards                      Drama  English  
Orange Is the New Black             Drama  English  
The Adventures of Puss in Boots  Children  English  
Dragons: Race to the Edge        Children  English  
Trolls: The Beat Goes On!        Children  English

In [ ]:
netflix[netflix['SeasonsParsed'] == 6] # the same, but in one line

Genre          GenreLabels  \
Title                                                                       
House of Cards                       Political drama      political,drama   
Orange Is the New Black                 Comedy-drama         comedy-drama   
The Adventures of Puss in Boots  childrens-animation  childrens-animation   
Dragons: Race to the Edge        childrens-animation  childrens-animation   
Trolls: The Beat Goes On!        childrens-animation  childrens-animation   

                                  Premiere                 Seasons  \
Title                                                                
House of Cards                    1-Feb-13  6 seasons, 73 episodes   
Orange Is the New Black          11-Jul-13  6 seasons, 78 episodes   
The Adventures of Puss in Boots  16-Jan-15  6 seasons, 78 episodes   
Dragons: Race to the Edge        26-Jun-15  6 seasons, 78 episodes   
Trolls: The Beat Goes On!        19-Jan-18  6 seasons, 38 episodes   

                                 SeasonsParsed  EpisodesParsed      Length  \
Title                                                                        
House of Cards                               6              73  42–59 min.   
Orange Is the New Black                      6              78  50–92 min.   
The Adventures of Puss in Boots              6              78  22–24 min.   
Dragons: Race to the Edge                    6              78  22–23 min.   
Trolls: The Beat Goes On!                    6              38  24–25 min.   

                                 MinLength  MaxLength   Status  Active  \
Title                                                                    
House of Cards                          42         59    Ended       0   
Orange Is the New Black                 50         92  Renewed       1   
The Adventures of Puss in Boots         22         24    Ended       0   
Dragons: Race to the Edge               22         23    Ended       0   
Trolls: The Beat Goes On!               24         25  Pending       1   

                                    Table Language  
Title                                               
House of Cards                      Drama  English  
Orange Is the New Black             Drama  English  
The Adventures of Puss in Boots  Children  English  
Dragons: Race to the Edge        Children  English  
Trolls: The Beat Goes On!        Children  English

In [ ]:
netflix[netflix['MaxLength'] > 100] # series longer than 100 minutes

Genre      GenreLabels  \
Title                                                                 
Sense8                             Science fiction  science-fiction   
Gilmore Girls: A Year in the Life     Family drama     family,drama   

                                    Premiere                 Seasons  \
Title                                                                  
Sense8                              5-Jun-15  2 seasons, 24 episodes   
Gilmore Girls: A Year in the Life  25-Nov-16              4 episodes   

                                   SeasonsParsed  EpisodesParsed       Length  \
Title                                                                           
Sense8                                         2              24  45–152 min.   
Gilmore Girls: A Year in the Life              0               4  88–102 min.   

                                   MinLength  MaxLength      Status  Active  \
Title                                                                         
Sense8                                    45        152       Ended       0   
Gilmore Girls: A Year in the Life         88        102  Miniseries       0   

                                   Table Language  
Title                                              
Sense8                             Drama  English  
Gilmore Girls: A Year in the Life  Drama  English

In [ ]:
# series with a number of episodes between 30 and 34
netflix[(netflix['EpisodesParsed'] >= 30) & (netflix['EpisodesParsed'] < 35)]

Genre      GenreLabels   Premiere  \
Title                                                             
Hemlock Grove       Horror/thriller  horror,thriller  19-Apr-13   
Bloodline                  Thriller         thriller  20-Mar-15   
Narcos                  Crime drama      crime,drama  28-Aug-15   
Love                Romantic comedy  romantic,comedy  19-Feb-16   
Santa Clarita Diet    Comedy-horror    comedy-horror   3-Feb-17   
Go! Live Your Way           Musical          musical  22-Feb-19   
Chef's Table           Culinary art     culinary,art  26-Apr-15   

                                   Seasons  SeasonsParsed  EpisodesParsed  \
Title                                                                       
Hemlock Grove       3 seasons, 33 episodes              3              33   
Bloodline           3 seasons, 33 episodes              3              33   
Narcos              3 seasons, 30 episodes              3              30   
Love                3 seasons, 34 episodes              3              34   
Santa Clarita Diet  3 seasons, 30 episodes              3              30   
Go! Live Your Way   2 seasons, 30 episodes              2              30   
Chef's Table        6 volumes, 30 episodes              0              30   

                        Length  MinLength  MaxLength   Status  Active  \
Title                                                                   
Hemlock Grove       45–58 min.         45         58    Ended       0   
Bloodline           48–68 min.         48         68    Ended       0   
Narcos              43–60 min.         43         60    Ended       0   
Love                24–40 min.         24         40    Ended       0   
Santa Clarita Diet  26–35 min.         26         35    Ended       0   
Go! Live Your Way   36–44 min.         36         44  Pending       1   
Chef's Table        42–58 min.         42         58  Renewed       1   

                              Table Language  
Title                                         
Hemlock Grove                 Drama  English  
Bloodline                     Drama  English  
Narcos                        Drama  English  
Love                         Comedy  English  
Santa Clarita Diet           Comedy  English  
Go! Live Your Way   ForeignLanguage  Spanish  
Chef's Table             DocuSeries  English

In [ ]:
# with 33 or 73 episodes
netflix[(netflix['EpisodesParsed'] == 33) | (netflix['EpisodesParsed'] == 73)]

Genre      GenreLabels   Premiere  \
Title                                                         
House of Cards  Political drama  political,drama   1-Feb-13   
Hemlock Grove   Horror/thriller  horror,thriller  19-Apr-13   
Bloodline              Thriller         thriller  20-Mar-15   

                               Seasons  SeasonsParsed  EpisodesParsed  \
Title                                                                   
House of Cards  6 seasons, 73 episodes              6              73   
Hemlock Grove   3 seasons, 33 episodes              3              33   
Bloodline       3 seasons, 33 episodes              3              33   

                    Length  MinLength  MaxLength Status  Active  Table  \
Title                                                                    
House of Cards  42–59 min.         42         59  Ended       0  Drama   
Hemlock Grove   45–58 min.         45         58  Ended       0  Drama   
Bloodline       48–68 min.         48         68  Ended       0  Drama   

               Language  
Title                    
House of Cards  English  
Hemlock Grove   English  
Bloodline       English

###  Sorting Rows

Let's create a new dataframe called `variable_length_shows` that contains series where longest episode is 45 minutes longer than the shortest

In [ ]:
variable_length_shows = netflix[netflix['MaxLength'] > netflix['MinLength'] + 45]
variable_length_shows

Genre          GenreLabels   Premiere  \
Title                                                                         
Sense8                      Science fiction      science-fiction   5-Jun-15   
Kong: King of the Apes  childrens-animation  childrens-animation  15-Apr-16   
Club de Cuervos                Comedy-drama         comedy-drama   7-Aug-15   

                                       Seasons  SeasonsParsed  EpisodesParsed  \
Title                                                                           
Sense8                  2 seasons, 24 episodes              2              24   
Kong: King of the Apes  2 seasons, 23 episodes              2              23   
Club de Cuervos         4 seasons, 45 episodes              4              45   

                             Length  MinLength  MaxLength Status  Active  \
Title                                                                      
Sense8                  45–152 min.         45        152  Ended       0   
Kong: King of the Apes   22–85 min.         22         85  Ended       0   
Club de Cuervos          36–94 min.         36         94  Ended       0   

                                  Table Language  
Title                                             
Sense8                            Drama  English  
Kong: King of the Apes         Children  English  
Club de Cuervos         ForeignLanguage  Spanish

We can use the `sort_values` method to sort on a chosen variable.

In [ ]:
variable_length_shows.sort_values('MaxLength')

Genre          GenreLabels   Premiere  \
Title                                                                         
Kong: King of the Apes  childrens-animation  childrens-animation  15-Apr-16   
Club de Cuervos                Comedy-drama         comedy-drama   7-Aug-15   
Sense8                      Science fiction      science-fiction   5-Jun-15   

                                       Seasons  SeasonsParsed  EpisodesParsed  \
Title                                                                           
Kong: King of the Apes  2 seasons, 23 episodes              2              23   
Club de Cuervos         4 seasons, 45 episodes              4              45   
Sense8                  2 seasons, 24 episodes              2              24   

                             Length  MinLength  MaxLength Status  Active  \
Title                                                                      
Kong: King of the Apes   22–85 min.         22         85  Ended       0   
Club de Cuervos          36–94 min.         36         94  Ended       0   
Sense8                  45–152 min.         45        152  Ended       0   

                                  Table Language  
Title                                             
Kong: King of the Apes         Children  English  
Club de Cuervos         ForeignLanguage  Spanish  
Sense8                            Drama  English

We can also sort in reverse (descending) order using `ascending=False`.

In [ ]:
variable_length_shows.sort_values('MaxLength', ascending=False)

Genre          GenreLabels   Premiere  \
Title                                                                         
Sense8                      Science fiction      science-fiction   5-Jun-15   
Club de Cuervos                Comedy-drama         comedy-drama   7-Aug-15   
Kong: King of the Apes  childrens-animation  childrens-animation  15-Apr-16   

                                       Seasons  SeasonsParsed  EpisodesParsed  \
Title                                                                           
Sense8                  2 seasons, 24 episodes              2              24   
Club de Cuervos         4 seasons, 45 episodes              4              45   
Kong: King of the Apes  2 seasons, 23 episodes              2              23   

                             Length  MinLength  MaxLength Status  Active  \
Title                                                                      
Sense8                  45–152 min.         45        152  Ended       0   
Club de Cuervos          36–94 min.         36         94  Ended       0   
Kong: King of the Apes   22–85 min.         22         85  Ended       0   

                                  Table Language  
Title                                             
Sense8                            Drama  English  
Club de Cuervos         ForeignLanguage  Spanish  
Kong: King of the Apes         Children  English

Note that `sort_values` does **not** actually modify the contents of the data frame!

If we use the ascendig order, we will see it is not sorted!

In [ ]:
variable_length_shows.sort_values('MaxLength')
variable_length_shows

Genre          GenreLabels   Premiere  \
Title                                                                         
Sense8                      Science fiction      science-fiction   5-Jun-15   
Kong: King of the Apes  childrens-animation  childrens-animation  15-Apr-16   
Club de Cuervos                Comedy-drama         comedy-drama   7-Aug-15   

                                       Seasons  SeasonsParsed  EpisodesParsed  \
Title                                                                           
Sense8                  2 seasons, 24 episodes              2              24   
Kong: King of the Apes  2 seasons, 23 episodes              2              23   
Club de Cuervos         4 seasons, 45 episodes              4              45   

                             Length  MinLength  MaxLength Status  Active  \
Title                                                                      
Sense8                  45–152 min.         45        152  Ended       0   
Kong: King of the Apes   22–85 min.         22         85  Ended       0   
Club de Cuervos          36–94 min.         36         94  Ended       0   

                                  Table Language  
Title                                             
Sense8                            Drama  English  
Kong: King of the Apes         Children  English  
Club de Cuervos         ForeignLanguage  Spanish

Solution: Create a new variable!

In [ ]:
new = variable_length_shows.sort_values('MaxLength')
new

Genre          GenreLabels   Premiere  \
Title                                                                         
Kong: King of the Apes  childrens-animation  childrens-animation  15-Apr-16   
Club de Cuervos                Comedy-drama         comedy-drama   7-Aug-15   
Sense8                      Science fiction      science-fiction   5-Jun-15   

                                       Seasons  SeasonsParsed  EpisodesParsed  \
Title                                                                           
Kong: King of the Apes  2 seasons, 23 episodes              2              23   
Club de Cuervos         4 seasons, 45 episodes              4              45   
Sense8                  2 seasons, 24 episodes              2              24   

                             Length  MinLength  MaxLength Status  Active  \
Title                                                                      
Kong: King of the Apes   22–85 min.         22         85  Ended       0   
Club de Cuervos          36–94 min.         36         94  Ended       0   
Sense8                  45–152 min.         45        152  Ended       0   

                                  Table Language  
Title                                             
Kong: King of the Apes         Children  English  
Club de Cuervos         ForeignLanguage  Spanish  
Sense8                            Drama  English

Creating a new dataset only with certain columns

In [ ]:
shorter_netflix = netflix[['Genre', 'SeasonsParsed']]
shorter_netflix

Genre  SeasonsParsed
Title                                                                     
House of Cards                              Political drama              6
Hemlock Grove                               Horror/thriller              3
Orange Is the New Black                        Comedy-drama              6
Marco Polo                                 Historical drama              2
Bloodline                                          Thriller              3
...                                                     ...            ...
Busted!                        Korean language variety show              1
The Break with Michelle Wolf                     Late-night              1
Norm Macdonald Has a Show                         Talk show              1
Patriot Act with Hasan Minhaj                     Talk show              0
The Fix                                          Panel show              1

[358 rows x 2 columns]

Adding new columns to a data frame (we will continue discussing this on the next Topic)

In [ ]:
# creating a new column named 'watched' filled with no
netflix['Watched'] = ['No']*len(netflix)
netflix

Genre  \
Title                                                         
House of Cards                              Political drama   
Hemlock Grove                               Horror/thriller   
Orange Is the New Black                        Comedy-drama   
Marco Polo                                 Historical drama   
Bloodline                                          Thriller   
...                                                     ...   
Busted!                        Korean language variety show   
The Break with Michelle Wolf                     Late-night   
Norm Macdonald Has a Show                         Talk show   
Patriot Act with Hasan Minhaj                     Talk show   
The Fix                                          Panel show   

                                                GenreLabels   Premiere  \
Title                                                                    
House of Cards                              political,drama   1-Feb-13   
Hemlock Grove                               horror,thriller  19-Apr-13   
Orange Is the New Black                        comedy-drama  11-Jul-13   
Marco Polo                                 historical,drama  12-Dec-14   
Bloodline                                          thriller  20-Mar-15   
...                                                     ...        ...   
Busted!                        korean,language,variety,show   4-May-18   
The Break with Michelle Wolf                     late-night  27-May-18   
Norm Macdonald Has a Show                         talk,show  14-Sep-18   
Patriot Act with Hasan Minhaj                     talk,show  28-Oct-18   
The Fix                                          panel,show  14-Dec-18   

                                              Seasons  SeasonsParsed  \
Title                                                                  
House of Cards                 6 seasons, 73 episodes              6   
Hemlock Grove                  3 seasons, 33 episodes              3   
Orange Is the New Black        6 seasons, 78 episodes              6   
Marco Polo                     2 seasons, 20 episodes              2   
Bloodline                      3 seasons, 33 episodes              3   
...                                               ...            ...   
Busted!                         1 season, 10 episodes              1   
The Break with Michelle Wolf    1 season, 10 episodes              1   
Norm Macdonald Has a Show       1 season, 10 episodes              1   
Patriot Act with Hasan Minhaj  3 volumes, 19 episodes              0   
The Fix                         1 season, 10 episodes              1   

                               EpisodesParsed      Length  MinLength  \
Title                                                                  
House of Cards                             73  42–59 min.         42   
Hemlock Grove                              33  45–58 min.         45   
Orange Is the New Black                    78  50–92 min.         50   
Marco Polo                                 20  48–65 min.         48   
Bloodline                                  33  48–68 min.         48   
...                                       ...         ...        ...   
Busted!                                    10  75–98 min.         75   
The Break with Michelle Wolf               10     27 min.         27   
Norm Macdonald Has a Show                  10  26–35 min.         26   
Patriot Act with Hasan Minhaj              19  23–30 min.         23   
The Fix                                    10  23–31 min.         23   

                               MaxLength   Status  Active        Table  \
Title                                                                    
House of Cards                        59    Ended       0        Drama   
Hemlock Grove                         58    Ended       0        Drama   
Orange Is the New Black               92  Renewed       1        Drama   
Marco Polo                 

## Finally
Try **harnessing the Power of AI** in Google Colab

* Generating code (e.g. remove all columns except Watched)
* View recommended plots (histograms, scatters, comparisons between data)
* New interactive sheet (asks for access to lots of things, runs better in Chrome, in Firefox it behaves weird)
* Convert to interactive table and suggest charts from there, clear output to return

## Lab Activity

If you are going to use Python, open `CM4125_Lab5.ipynb` using **Google Colab** and follow the steps.

Or, you can read `CM4125_Lab5_guided.html` to follow the logic used on the data.
    
* Try to replicate the steps in your preferred tool!